### PR TITLE
Framework: move z-index values to z-index scss function

### LINKED ISSUE #1139: Framework: Use named SCSS variables for z-index values
We could create variables like this:

```
$bottom-layer: 0;
$middle-layer: 1;
$overlay-layer: 9999;
```

If we don't have some kind of system for this, we're going to end up with random z-indexes scattered throughout the code:

````
grep -orP --no-filename "z-index:[^;]*" stylesheets | sort | uniq -c

      8 z-index: 0
     17 z-index: 1
      5 z-index: 10
      6 z-index: 1000
      2 z-index: 10000
      1 z-index: 100000
      1 z-index: 11
      2 z-index: 11111
      2 z-index: 18
      1 z-index: 180
      1 z-index: 19
      8 z-index: 2
      5 z-index: 20
      1 z-index: 20000
      1 z-index: 201
      4 z-index: 3
      1 z-index: 30000
      1 z-index: 4000
      1 z-index: 40000
      1 z-index: 500
      1 z-index: 800
      2 z-index: 900
      3 z-index: 999
      1 z-index: 9998
      5 z-index: 9999
      1 z-index: 99999
      1 z-index: auto
````

### PR DESCRIPTION
**Pull request by gwwar**
Fixes #1139. This PR tidies up our static z-index values, and makes it easy to check against our different stacking contexts.

A future PR will address components like popover that need dynamically created z-indexes, because they are inserted into a variety of contexts.

Also see: [sass functions](http://thesassway.com/advanced/pure-sass-functions)

Before:
```scss
.environment-badge { 
    z-index: 999;
}
```

After:
```scss
.environment-badge { 
    z-index: z-index( 'root', '.environment-badge' );
}
```

When adding z-indexes use our scss z-index function. This means you'll need to add
another entry to the `$z-layers` variable in `assets/stylesheets/shared/functions/_z-index.scss`

Because browsers support a hierarchy of stacking contexts rather than a single global one, we use two keys: the name of the parent stacking context, and a unique name for the stacking context you've just created. We do this so we can quickly compare z-index values within a given stacking context. 

A stacking context is created when:

1. It's the root element (HTML)
2. Has a position other than static, with a z-index value
3. position:fixed
4. Has one of the following css properties: (transform, opacity<1, mix-blend-mode, filter)
5. isolation:isolate
6: -webkit-overflow-scrolling: touch

So, to add a new z-index:

1. You'll want to make sure the element actually creates a stacking context
2. Lookup what it's parent stacking context
3. Put the new rule in the $z-layers tree, using the stacking context(s) as the initial key(s) and the element's full CSS selector as the final key.

You can run this handy [gist](https://gist.github.com/gwwar/2f661deec7b99a1a418b) in the console to find both.

```scss
$z-layers: (
    'root': (
        '.environment-badge': 999,
        '.some-other-stacking-context': 10000
    ),
    '.environment-badge': (
        '.environment-badge .environment::before': -1,
        '.environment-badge .bug-report': 1000
    )
);
```

### Testing

The diff between style.css in this branch and in master should only contain the removal of a duplicate z-index value.

### CODE REVIEW

**Comment by gwwar** [comment]
cc @adambbecker @aduth @rralian

**Comment by aduth** [comment]
Can you explain the advantage of having a tree, rather than a flattened list? It's not clear to me when I'd want to add a `z-index` to, for example, `content` instead of `body`. I can see how it might help with grouping in the relative case, though we could potentially apply a naming convention there similar to or matching our class naming syntax (e.g. `z-index( 'sharing-buttons__more' );`). I just worry that it'll be daunting for people to interact with the tree, since I might not know when I should add a new base key, and having a sub-tree with a single item occupies three total lines in the file.

Let me also reiterate that this is awesome, and a much-needed move away from our current wild-west approach (`z-index + 1` :smile: ).

**Comment by rralian** [comment]
I see good documentation in comments in [_z-index.scss](https://github.com/Automattic/wp-calypso/pull/1392/files#diff-303494311390095357dc6a38984e0cc9), but can we also get that into something more discoverable? It's probably worth a section on our [css guidelines doc](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md), or if the explanation winds up getting too large, a link from there to a separate markdown doc (maybe `css__z-index.md`)?

> Can you explain the advantage of having a tree, rather than a flattened list? It's not clear to me when I'd want to add a z-index to, for example, content instead of body.

@aduth the use of a tree ties to the idea of stacking contexts, because z-indices around the page are not compared absolutely to each other, but only within their stacking context. So element A with a lower z-index could appear above element B with a higher z-index element because element A's parent has a higher z-index than element B's parent. But I agree that some more instruction would be helpful here. Maybe a couple of example cases... one where we add an element to an existing stacking context, and one where it would be appropriate to create a new base context. I think that would make sense in the markdown documentation I mention above.

**Comment by gwwar** [comment]
@rralian @aduth thanks for taking a look. :)

I can add additional examples/documentation as it looks like how z-index and render order works can be misinterpreted pretty easily. I'll try adding a few live examples too, since this is probably a bit difficult to explain without seeing it in action.

File: assets/stylesheets/shared/functions/_z-index.scss [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+	'secondary': (),
+	// under body / wpcom / layout / content
+	'tertiary': (),
+
+	//=================================================
+	// relative
+	//=================================================
+	'environment-badge': (
+			'before': -1
+	),
```
Comment (nylen) [inline assets/stylesheets/shared/functions/_z-index.scss]: ^ Extra tab here

**Comment by gwwar** [comment]
> we could potentially apply a naming convention there similar to or matching our class naming syntax

I'd be fine with using BEM-ish naming convention for the keys. I would in fact encourage folks to double check if they need z-index rules and rename keys if they have better names. 

We can also probably remove a number of rules in future PRs. There were a decent number of components that could have reordered elements to take advantage of the natural stacking order, or thought their elements were in the top stacking context but weren't really.

File: assets/stylesheets/shared/functions/_z-index.scss (hunk 1 above)
Comment (gwwar) [inline assets/stylesheets/shared/functions/_z-index.scss]: :eyes: :+1:

**Comment by nylen** [comment]
This is great :smile: 

I did a comparison of the `style.css` build before and after this change:  https://gist.github.com/nylen/d3354e3506350c356481

The only problem I see is that the generated CSS contains one of these `z-index` calls - I'm not really sure how that is possible though:

```diff
<snip>
   .screen-reader-text:focus, .masterbar-new-post__label:focus {
<snip>
-    z-index: 100000; }
+    z-index: z-index("screen-reader-text", "focus"); }
```

There were a few other things that changed.  I'm assuming these changes were intentional, but I'm going to run through and check them, and I'll update this comment when I've done that.

+1 on some documentation/examples.

**Comment by gwwar** [comment]
@nylen thanks for checking! I'll try to tidy that up. I removed two z-index rules that appeared to do nothing, but I may add them back in just so the PR is easier to review.

**Comment by gwwar** [comment]
@nylen the style.css diff should now just contain a removal of a duplicate z-index rule. documentation/examples will be ready shortly.

**Comment by rralian** [comment]
> I may add them back in just so the PR is easier to review

Yeah, that's always nice to move little fixes and refactoring to a separate PR. I'm not always good about that myself though.

**Comment by nylen** [comment]
Confirmed, the code looks good now :+1:

**Comment by gwwar** [comment]
I've added a gist to return the closest stacking context: https://gist.github.com/gwwar/2f661deec7b99a1a418b I'll do another passthrough on the z-index map tomorrow to make sure I've got the right parent contexts.

**Comment by gwwar** [comment]
@aduth @nylen @rralian Let me know if the stacking context explanation in css.md makes sense, if not I'll probably need to create a separate .md file for it.

File: client/auth/style.scss [hunk 2]
```diff
@@ -57,7 +57,7 @@ html.is-desktop {
 
 		.form-fieldset input {
 		position: relative;
-		z-index: 1;
+		z-index: z-index( 'root', 'auth__form__input' );
```
Comment (aduth) [inline client/auth/style.scss]: Can you explain what the naming convention is here for `auth__form__input` ? Is the idea to use double-underscore in place of a space in the selector? Should we prescribe this in the `css.md` docs?

Also, is there any reason we couldn't just use the full selector, spaces and all, in the string?

```sass
z-index: z-index( 'root', '.auth__form .form-fieldset input' );
```

**Comment by gwwar** [comment]
> Also, is there any reason we couldn't just use the full selector, spaces and all, in the string?

I wouldn't be opposed to that. Some of these were nested so deeply, bem naming broke. :) Any other folks have opinions on naming stacking contexts, before I dive into updating key names again?

**Comment by nylen** [comment]
Including the full selector as the key makes sense to me too.  We'll have to encourage people to keep the keys updated as the selectors change.

Made some tweaks to the documentation in 0ca5506 (mostly formatting).  Also rewrote it with full selectors as keys and added a 3rd step detailing how to add new entries to `$z-layers`.

The code is written so that we can have more than 2 levels of keys.  Is this something we expect and want to document?  I changed the wording in the docs to allow for that, but maybe that should be updated further.

**Comment by gwwar** [comment]
Thanks for the documentation tweaks! I'll update the keys today.

> The code is written so that we can have more than 2 levels of keys. Is this something we expect and want to document?

There isn't too much benefit to it in this case. I really just wanted to highlight that:
1. You might not be working in the root stacking context
2. Be able to quickly compare values within a stacking context
3. To remind folks that they're creating a new stacking context

**Comment by gwwar** [comment]
Heads up that I'll be squashing changes so rebasing is a bit easier :)

**Comment by gwwar** [comment]
@nylen @aduth keys updated in https://github.com/Automattic/wp-calypso/commit/7d1c388652afd205736cc790e8233529b3afe75d

**Comment by nylen** [comment]
Sorry I missed this update earlier today.  Verified the generated CSS again after the latest changes.

Thanks for making this & seeing it through.  Excellent work, let's :shipit:

**Comment by gwwar** [comment]
:tada: :cake:

**Comment by aduth** [comment]
Happy to see this merged :smiley: Now the interesting part - to ensure widespread adoption!

**Pull request merged**